### PR TITLE
[Files] Verfy a sharing link asap and give nice error 

### LINKED ISSUE #1671: Verify nonce right away when visiting a sharing link
We now have the nonce verification endpoint to prevent users from accessing a nonce that has been revoked.
We should use this to redirect a user to an error page.

### PR DESCRIPTION
**Pull request by Tbaut**
closes #1671

- verify the nonce asap if the user isn't logged-in
- if the user is already logged in, the redirect to the landing happens and a nice (translatable) error is displayed.

### CODE REVIEW

**Comment by asnaith** [comment]
Hey Thibaut, I found an issue whilst testing this out. When a logged in user clicks an invalid link they see the new error message but also other elements that should be hidden: 

Screen Shot 2021-10-27 at 7 41 35 PM

There's also a small discrepancy with the icon colour of the exclamation mark in dark mode when the user is signed out. 

Screen Shot 2021-10-27 at 7 43 37 PM

File: packages/files-ui/src/Components/Modules/LinkSharingModule.tsx [hunk 1]
```diff
@@ -56,10 +56,12 @@ const useStyles = makeStyles(
     })
 )
 
-interface DecodedJwt {
+export interface DecodedNounceJwt {
```
Comment (RyRy79261) [inline packages/files-ui/src/Components/Modules/LinkSharingModule.tsx]:
```suggestion
export interface DecodedNonceJwt {
```

File: packages/files-ui/src/Components/Modules/LinkSharingModule.tsx [hunk 2]
```diff
@@ -71,15 +73,26 @@ const LinkSharingModule = () => {
   const [encryptedEncryptionKey, setEncryptedEncryptionKey] = useState("")
   const [error, setError] = useState("")
   const classes = useStyles()
-  const { bucket_id: bucketId, permission } = useMemo(() => {
+  const { bucket_id: bucketId, permission, nonce_id } = useMemo(() => {
     try {
-      return (jwt && jwtDecode<DecodedJwt>(jwt)) || {}
+      return (jwt && jwtDecode<DecodedNounceJwt>(jwt)) || {}
```
Comment (RyRy79261) [inline packages/files-ui/src/Components/Modules/LinkSharingModule.tsx]:
```suggestion
      return (jwt && jwtDecode<DecodedNonceJwt>(jwt)) || {}
```

File: packages/files-ui/src/Components/Modules/LoginModule/InitialScreen.tsx [hunk 3]
```diff
@@ -23,6 +24,9 @@ import { IdentityProvider } from "@chainsafe/files-api-client"
 import PasswordlessEmail from "./PasswordlessEmail"
 import { Form, FormikProvider, useFormik } from "formik"
 import { emailValidation } from "../../../Utils/validationSchema"
+import { getJWT } from "../../../Utils/pathUtils"
+import jwtDecode from "jwt-decode"
+import { DecodedNounceJwt } from "../LinkSharingModule"
```
Comment (RyRy79261) [inline packages/files-ui/src/Components/Modules/LoginModule/InitialScreen.tsx]:
```suggestion
import { DecodedNonceJwt } from "../LinkSharingModule"
```

File: packages/files-ui/src/Components/Modules/LoginModule/InitialScreen.tsx [hunk 4]
```diff
@@ -186,6 +193,28 @@ const InitialScreen = ({ className }: IInitialScreen) => {
...
+  const [isValidNonce, setIsValidNonce] = useState(true)
+
+  useEffect(() => {
+    if (!isSharing) return
+
+    const jwt = getJWT(state?.from)
+    let nonce = ""
+
+    try {
+      nonce = (jwt && jwtDecode<DecodedNounceJwt>(jwt).nonce_id) || ""
```
Comment (RyRy79261) [inline packages/files-ui/src/Components/Modules/LoginModule/InitialScreen.tsx]:
```suggestion
      nonce = (jwt && jwtDecode<DecodedNonceJwt>(jwt).nonce_id) || ""
```

**Review by RyRy79261 — COMMENTED** [review]
Looks streamlined though spotted a typo :)

**Review by FSM1 — CHANGES_REQUESTED** [review]
Works well for revoked links. Would probably make sense to allow the user some way to go back to the normal login page if they attempt to join using a revoked link:
image

Another flow that could be improved is if the user provides a malformed JWT in the join link. Since its not possible to extract the NONCE from this JWT, it never gets validated, and only throws an error after the user logs in.

**Comment by asnaith** [comment]
Looking good now since the latest changes @Tbaut! 

Small observation with the new go back button - If I'm logged out, click a link, see invalid link message, then return to login screen I will see the "Sign in/up to access the shared folder" message and then if I sign in I see the invalid link message once more. Perhaps when a user clicks to go back we should direct them to the regular login URL to prevent this?

**Comment by Tbaut** [comment]
Thanks for the thorough review y'all. I did a lot of code changes, please re-check as much as you can.
- no more loader + invalid link info at the same time
- no more typo
- no more wrong color on dark mode
- there's a button to go to normal login now (if the user isn't logged in)
- nice error instead of doing nothing if the JWT is malformed

**Comment by Tbaut** [comment]
> Perhaps when a user clicks to go back we should direct them to the regular login URL to prevent this?

Arg that's annoying. Yeah, I'll redirect then.

edit: @asnaith should be sorted with 579f131

**Comment by asnaith** [comment]
I've thoroughly reviewed the feature and updated the user scenarios doc. Everything is working great 😃

✅ User can generate a link for read access to a share
✅ User can generate a link for edit access to a share
✅ User can delete a share link that they have previously created
✅ User can copy a generated link to their keyboard
✅ User is informed that they need to be logged in / signed up to access share
✅ User is informed of their access level on the "You were added" modal
✅ User can navigate to the share via a browse button on the "You were added" modal
✅ User joining a share via read rights link should have read only access
✅ User joining a share via edit rights link should have edit access
✅ Existing read rights user should have their access rights upgraded after clicking an edit rights link
✅ Existing edit rights user should not have their access rights downgraded after clicking a read rights link
✅ Signed in user can see invalid link error if share link was deleted
✅ Signed out user can see invalid link error if share link was deleted
✅ Signed out user can return to the login screen after clicking an invalid share link
✅ User can see "Token Malformed" error when clicking an incorrect share link

**Review by asnaith — APPROVED** [review]

**Review by FSM1 — APPROVED** [review]

**Pull request merged**